examples/research_projects/jax-projects/README.md

```diff
@@ -383,13 +383,305 @@ official [flax example folder](https://github.com/huggingface/transformers/tree/
+
+
+#### **How to write a training loop**
+
+Now let's see how we can write a simple training loop to train Flax models, we will use `FlaxGPT2ForCausalLM` as an example. 
+
+A training loop for Flax models typically consists of
+- A loss function that takes the parameters and inputs, runs the forward pass and returns the loss. 
+- We then transform the loss function using `jax.grad` or `jax.value_and_grad`  so that we get the gradients of all parameters.
+- An optimizer to update the paramteres using the gradients returned by the transformed loss function.
+- A train step function which combines the loss function and optimizer update, does the forward and backward pass and returns the updated parameters.
+
+Lets see how that looks like in code:
+
+First initialize our model
+
+```python
+import jax
+import jax.numpy as jnp
+
+from transformers import FlaxGPT2ForCausalLM
+
+model = FlaxGPT2ForCausalLM(config) 
+```
+
+As explained above we don't compute the loss inside the model, so we will define the loss
+function here.
+
+```python
+from flax.training.common_utils import onehot
+
+def cross_entropy(logits, labels):
+   return -jnp.sum(labels * jax.nn.log_softmax(logits, axis=-1), axis=-1)
+
+# define a function which will run the forward pass return loss
+def compute_loss(params, input_ids, labels):
+   logits = model(input_ids, params=params, train=True)
+   loss = cross_entropy(logits, onehot(labels)).mean()
+   return loss
+```
 
-TODO (should be filled by 29.06.)
+Now we transform the loss function with `jax.value_and_grad`.
+
+```python
+# transform the loss function to get the gradients
+grad_fn = jax.value_and_grad(compute_loss)
+```
+
+Initialize our optimizer, we use the [optax](https://github.com/deepmind/optax) library to get the optimizer. 
+
+```python
+import optax
+
+params = model.params
+tx = optax.sgd(learning_rate=3e-3)
+opt_state = tx.init(params)
+```
+
+Now we define a single training step which will do a forward and a backward pass.
+
+```python
+def _train_step(params, opt_state, input_ids, labels)
+   # do the forward pass and get the loss and gradients
+   loss, grads = grad_fn(params, input_ids, labels)
+
+   # use the gradients to update parameters
+   updates, opt_state = tx.update(grads, opt_state)
+   updated_params = optax.apply_updates(params, updates)
+
+   return updates_params, opt_state, loss
+
+train_step = jax.jit(_train_step)
+```
+
+Finally, let's run our training loop.
+
+```python
+# train loop
+for i in range(10):
+   params, opt_state, loss = train_step(params, opt_state, input_ids, labels)
+```
+
+Note how we always pass the `params` and `opt_state` to the `train_step` which then returns the updated `params` and `opt_state`. This is becuase of the staless nature of JAX/Flax models, all the state
```

Comment:
```suggestion
Note how we always pass the `params` and `opt_state` to the `train_step` which then returns the updated `params` and `opt_state`. This is because of the staless nature of JAX/Flax models, all the state
```